### PR TITLE
Use HMAC-based CSRF tokens

### PR DESCRIPTION
**Pull request by ThrawnCA**
This will protect against the parent/sub-domain attacks described at https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md#double-submit-cookie

### CODE REVIEW

File: ckanext/qgov/common/anti_csrf.py [hunk 1]
```diff
@@ -131,11 +176,11 @@ def is_request_exempt():
     return not is_logged_in() or API_URL.match(request.path) or request.method in {'GET', 'HEAD', 'OPTIONS'}
 
 def anti_csrf_before(obj, action, **params):
+    RAW_BEFORE(obj, action)
```
Comment (duttonw) [inline ckanext/qgov/common/anti_csrf.py]: why was this moved?

File: ckanext/qgov/common/anti_csrf.py [hunk 2]
```diff
@@ -104,25 +122,52 @@ def get_response_token():
...
+
+    return config.get('beaker.session.secret')
+
 def create_response_token():
-    import binascii, os
-    token = binascii.hexlify(os.urandom(32))
+    import time, random, hmac, hashlib, urllib
+
+    secret_key = _get_secret_key()
+    username = urllib.quote(g.userobj.name)
```
Comment (duttonw) [inline ckanext/qgov/common/anti_csrf.py]: you should drop "/" from username if it does have one.

File: ckanext/qgov/common/anti_csrf.py [hunk 3]
```diff
@@ -94,6 +83,35 @@ def get_cookie_token():
...
+    message = parts[1]
+
+    secret_key = _get_secret_key()
+    expected_hmac = unicode(hmac.HMAC(secret_key, message, hashlib.sha512).hexdigest())
+    if not hmac.compare_digest(expected_hmac, provided_hmac):
+        return False
+
+    message_parts = message.split('/', 2)
+    timestamp = int(message_parts[0])
+    username = message_parts[2]
```
Comment (duttonw) [inline ckanext/qgov/common/anti_csrf.py]: you should drop / if it has it to match generator. better yet, have a method that generates the message, and a method that spits out an array of what was in the message thats standard.

File: ckanext/qgov/common/anti_csrf.py (hunk 1 above)
Comment (ThrawnCA) [reply ckanext/qgov/common/anti_csrf.py]: The built-in `before` function initialises some necessary objects such as the current user.

File: ckanext/qgov/common/anti_csrf.py (hunk 2 above)
Comment (ThrawnCA) [reply ckanext/qgov/common/anti_csrf.py]: Why would a username contain `/`?
